--- linux-4.15/tools/bpf/bpf_jit_disasm.c.orig	2017-09-03 22:56:17.000000000 +0200
+++ linux-4.15/tools/bpf/bpf_jit_disasm.c	2017-11-03 22:34:58.128950311 +0100
@@ -71,7 +71,7 @@
 
 	disassemble_init_for_target(&info);
 
-	disassemble = disassembler(bfdf);
+	disassemble = disassembler(info.arch, bfd_big_endian(bfdf), info.mach, bfdf);
 	assert(disassemble);
 
 	do {
--- linux-4.15/tools/bpf/bpftool/jit_disasm.c.orig	2018-01-28 22:20:33.000000000 +0100
+++ linux-4.15/tools/bpf/bpftool/jit_disasm.c	2018-02-07 21:40:01.920963503 +0100
@@ -107,7 +107,7 @@
 
 	disassemble_init_for_target(&info);
 
-	disassemble = disassembler(bfdf);
+	disassemble = disassembler(info.arch, bfd_big_endian(bfdf), info.mach, bfdf);
 	assert(disassemble);
 
 	if (json_output)
