--- linux-5.7/tools/lib/traceevent/plugins/Makefile.orig	2020-06-01 01:49:15.000000000 +0200
+++ linux-5.7/tools/lib/traceevent/plugins/Makefile	2020-07-27 17:00:03.592989940 +0200
@@ -197,7 +197,7 @@
 	xargs echo "U w W" | tr 'w ' 'W\n' | sort -u | xargs echo`;\
 	if [ "$$symbol_type" = "U W" ];then				\
 		(echo '{';                                              \
-		$(NM) -u -D $1 | awk 'NF>1 {print "\t"$$2";"}' | sort -u;\
+		$(NM) -u -D $1 | awk 'NF>1 {gsub("@.*", "", $$2); print "\t"$$2";"}' | sort -u;\
 		echo '};';                                              \
 		) > $2;                                                 \
 	else                                                            \
