--- linux-4.12/tools/lguest/lguest.c.orig	2017-07-03 01:07:02.000000000 +0200
+++ linux-4.12/tools/lguest/lguest.c	2017-07-13 17:27:08.446483634 +0200
@@ -77,6 +77,7 @@
 #include <linux/virtio_ring.h>
 #include "../../include/uapi/linux/virtio_pci.h"
 #include <asm/bootparam.h>
+#include <asm/e820.h>
 #include "../../include/linux/lguest_launcher.h"
 
 #define BRIDGE_PFX "bridge:"
@@ -3339,7 +3340,7 @@
 	 * simple, single region.
 	 */
 	boot->e820_entries = 1;
-	boot->e820_table[0] = ((struct e820_entry) { 0, mem, E820_TYPE_RAM });
+	boot->e820_table[0] = ((struct boot_e820_entry) { 0, mem, E820_RAM });
 	/*
 	 * The boot header contains a command line pointer: we put the command
 	 * line after the boot header.
