--- linux-5.3/tools/perf/trace/beauty/usbdevfs_ioctl.sh.orig	2019-09-15 23:19:32.000000000 +0200
+++ linux-5.3/tools/perf/trace/beauty/usbdevfs_ioctl.sh	2019-09-24 18:22:01.021863039 +0200
@@ -15,7 +15,7 @@
 printf "#if 0\n"
 printf "static const char *usbdevfs_ioctl_32_cmds[] = {\n"
 regex="^#[[:space:]]*define[[:space:]]+USBDEVFS_(\w+)[[:space:]]+_IO[WR]{0,2}\([[:space:]]*'U'[[:space:]]*,[[:space:]]*([[:digit:]]+).*"
-egrep $regex ${header_dir}/usbdevice_fs.h | egrep 'USBDEVFS_\w+32[[:space:]]' | \
+egrep "$regex" ${header_dir}/usbdevice_fs.h | egrep 'USBDEVFS_\w+32[[:space:]]' | \
 	sed -r "s/$regex/\2 \1/g"	| \
 	sort | xargs printf "\t[%s] = \"%s\",\n"
 printf "};\n"
